{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import gpcrdb\n",
    "import matplotlib\n",
    "matplotlib.rc('pdf', fonttype=42)\n",
    "matplotlib.use('Agg')\n",
    "from matplotlib import patches\n",
    "import matplotlib.pyplot as plt\n",
    "plt.rcParams['font.family'] = \"Arial\"\n",
    "from utils import VariationType\n",
    "import numpy as np\n",
    "import ensembl\n",
    "import json\n",
    "from config import AM_THRESHOLD_BENIGN, AM_THRESHOLD_PATHOGENIC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "jpn_vars = {}\n",
    "global_vars = {}\n",
    "for receptor in gpcrdb.get_filtered_receptor_list():\n",
    "    with open(receptor.ensembl_path) as f:\n",
    "        display_name = json.load(f)['display_name']\n",
    "    with open(receptor.japan_cds_csv_path) as f:\n",
    "        for l in f.readlines():\n",
    "            try:\n",
    "                anno = ensembl.Annotation.from_csv_line(l)                    \n",
    "            except ensembl.BlankLineError:\n",
    "                continue\n",
    "\n",
    "            if anno.var_type != VariationType.MISSENSE:\n",
    "                continue\n",
    "\n",
    "            if not anno.pathogenicity:\n",
    "                continue\n",
    "\n",
    "            jpn_vars[(display_name, anno.snv.chromosome, anno.snv.position, anno.snv.ref, anno.snv.alt)] = anno\n",
    "\n",
    "    with open(receptor.global_cds_csv_path) as f:\n",
    "        for l in f.readlines():\n",
    "            try:\n",
    "                anno = ensembl.Annotation.from_csv_line(l)                    \n",
    "            except ensembl.BlankLineError:\n",
    "                continue\n",
    "\n",
    "            if anno.var_type != VariationType.MISSENSE:\n",
    "                continue\n",
    "\n",
    "            if not anno.pathogenicity:\n",
    "                continue\n",
    "\n",
    "            global_vars[(display_name, anno.snv.chromosome, anno.snv.position, anno.snv.ref, anno.snv.alt)] = anno\n",
    "\n",
    "all_var_keys = set(jpn_vars.keys()) | set(global_vars.keys())\n",
    "pathos = {k: jpn_vars[k].pathogenicity if k in jpn_vars.keys() else global_vars[k].pathogenicity for k in all_var_keys}\n",
    "\n",
    "likely_pathogenic_keys = tuple(k for k, v in pathos.items() if AM_THRESHOLD_PATHOGENIC < v)\n",
    "ambigous_keys = tuple(k for k, v in pathos.items() if AM_THRESHOLD_BENIGN <= v <= AM_THRESHOLD_PATHOGENIC)\n",
    "likely_benign_keys = tuple(k for k, v in pathos.items() if v < AM_THRESHOLD_BENIGN)\n",
    "\n",
    "likely_pathogenic_freqs = {k: (jpn_vars[k].snv.AF if k in jpn_vars.keys() else 0, global_vars[k].snv.AF if k in global_vars.keys() else 0) for k in likely_pathogenic_keys}\n",
    "ambigous_freqs = {k: (jpn_vars[k].snv.AF if k in jpn_vars.keys() else 0, global_vars[k].snv.AF if k in global_vars.keys() else 0) for k in ambigous_keys}\n",
    "likely_benign_freqs = {k: (jpn_vars[k].snv.AF if k in jpn_vars.keys() else 0, global_vars[k].snv.AF if k in global_vars.keys() else 0) for k in likely_benign_keys}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "meta NOT subset; don't know how to subset; dropped\n"
     ]
    }
   ],
   "source": [
    "# For Fig. 4a\n",
    "fig, ax = plt.subplots(1, 1, figsize=(5, 5), dpi=300)\n",
    "\n",
    "ax.fill_between([-1, 2], [-0.75, 2.25], [-1.25, 1.75], color='lightgray', lw=0, alpha=0.8, zorder=-1000)\n",
    "ax.fill_between([-1, 2], [-0.5, 2.5], [-1.5, 1.5], color='lightgray', lw=0, alpha=0.6, zorder=-900)\n",
    "ax.fill_between([-1, 2], [-0.25, 2.75], [-1.75, 1.25], color='lightgray', lw=0, alpha=0.4, zorder=-800)\n",
    "\n",
    "# Plot only AF > 0.25\n",
    "frequent_likely_benign_freqs = [freqs for freqs in likely_benign_freqs.values() if freqs[0] > 0.25 or freqs[1] > 0.25]\n",
    "frequent_ambigous_freqs = [freqs for freqs in ambigous_freqs.values() if freqs[0] > 0.25 or freqs[1] > 0.25]\n",
    "frequent_likely_pathogenic_freqs = [freqs for freqs in likely_pathogenic_freqs.values() if freqs[0] > 0.25 or freqs[1] > 0.25]\n",
    "ax.scatter(*np.array(frequent_likely_benign_freqs).T, marker='.', color='tab:gray', label=\"likely benign\", zorder=1)\n",
    "ax.scatter(*np.array(frequent_ambigous_freqs).T, marker='.', color='tab:orange', label=\"ambigous\", zorder=10)\n",
    "ax.scatter(*np.array(frequent_likely_pathogenic_freqs).T, marker='.', color='tab:red', label=\"likely pathogenic\", zorder=100)\n",
    "\n",
    "# AF <= 0.25 will be plotted in another fig\n",
    "num_infrequent = len(all_var_keys) - len(frequent_ambigous_freqs) - len(frequent_likely_benign_freqs) - len(frequent_likely_pathogenic_freqs)\n",
    "ax.fill_between([0, 0.25], [0, 0], [0.25, 0.25], color='white', lw=0, zorder=-600)\n",
    "ax.text(0.125, 0.125, \"Fig. S4a\\n{:,} vars.\".format(num_infrequent), va='center', ha='center', color='tab:gray', zorder=-500)\n",
    "ax.text(0, 1, \"Total {:,} vars. (incl. AF$\\leq$0.25)\".format(len(all_var_keys)), va='top', ha='left', color='tab:gray', zorder=-500)\n",
    "\n",
    "# Japanese-specific variants\n",
    "jpn_freq_threshold, global_freq_threshold = 0.25, 0.015\n",
    "x, y = jpn_freq_threshold, -global_freq_threshold\n",
    "w, h = 1 - jpn_freq_threshold + 0.015, global_freq_threshold * 2\n",
    "r = patches.Rectangle(xy=(x, y), width=w, height=h, ec='tab:gray', fill=False, zorder=-1, lw=0.6)\n",
    "ax.add_patch(r)\n",
    "ax.text(1, global_freq_threshold, \"Table S1\", color='tab:gray', ha='right', va='bottom')\n",
    "\n",
    "lines = []\n",
    "for k in all_var_keys:\n",
    "    jpn_var, global_var = jpn_vars.get(k, None), global_vars.get(k, None)\n",
    "    jpn_freq, global_freq = jpn_var.snv.AF if jpn_var else 0, global_var.snv.AF if global_var else 0\n",
    "    if jpn_freq > jpn_freq_threshold and global_freq < global_freq_threshold:\n",
    "        display_name = k[0]\n",
    "        superscript = jpn_var.generic_number if jpn_var.generic_number else jpn_var.segment\n",
    "        substitution = \"{}{}^{{{}}}{}\".format(jpn_var.ref_aa, jpn_var.residue_number, superscript, jpn_var.alt_aa)\n",
    "        cols = [display_name, substitution, jpn_var.snv.rsid, jpn_freq, global_freq]\n",
    "        lines.append(cols)\n",
    "lines.sort(key=lambda l: float(l[3]), reverse=True)\n",
    "with open(\"./figures/TableS1_japanese_specifics.csv\", 'w') as f:\n",
    "    f.write(\",\".join([\"#GPCR\", \"Substitution\", \"rsID\", \"AF_jpn\", \"AF_global\"]) + \"\\n\")\n",
    "    f.write('\\n'.join([\",\".join([str(v) for v in l]) for l in lines]))\n",
    "\n",
    "notables = []\n",
    "for k in likely_pathogenic_keys + ambigous_keys:\n",
    "    freqs = (jpn_vars[k].snv.AF if k in jpn_vars.keys() else 0, global_vars[k].snv.AF if k in global_vars.keys() else 0)\n",
    "    if freqs[0] > 0.25 or freqs[1] > 0.25:\n",
    "        anno: ensembl.Annotation = jpn_vars.get(k, global_vars[k])\n",
    "        display_name = k[0]\n",
    "        superscript = anno.generic_number if anno.generic_number else anno.segment\n",
    "        substitution = \"{}{}$^{{{}}}${}\".format(anno.ref_aa, anno.residue_number, superscript, anno.alt_aa)\n",
    "        notables.append([display_name, substitution, anno.snv.rsid, anno.pathogenicity, *freqs])\n",
    "notables.sort(key=lambda n: n[5], reverse=True)\n",
    "with open(\"./figures/4a_notables.csv\", 'w') as f:\n",
    "    f.write('\\n'.join([','.join([str(v) for v in l]) for l in notables]))\n",
    "\n",
    "ax.legend(loc='lower center', bbox_to_anchor=(0.5, 1), ncol=3)\n",
    "lims = (-0.02, 1.02)\n",
    "ticks = [0, 0.25, 0.5, 0.75, 1]\n",
    "ax.set_xlim(*lims)\n",
    "ax.set_xticks(ticks)\n",
    "ax.set_xticklabels(ticks)\n",
    "ax.set_yticks(ticks)\n",
    "ax.set_yticklabels(ticks)\n",
    "ax.set_ylim(*lims)\n",
    "ax.set_aspect('equal')\n",
    "ax.set_xlabel(\"Allele Freq. in 54KJPN\")\n",
    "ax.set_ylabel(\"Allele Freq. in 1KGP\")\n",
    "fig.tight_layout()\n",
    "fig.savefig(\"./figures/4a_jpn_vs_global.pdf\")\n",
    "plt.close(fig)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Fig. S4a\n",
    "fig, ax = plt.subplots(1, 1, figsize=(5, 5), dpi=300)\n",
    "ax.set_facecolor(color='#d4d4d4')\n",
    "\n",
    "# Plot only AF <= 0.25\n",
    "infrequent_likely_benign_freqs = [freqs for freqs in likely_benign_freqs.values() if freqs[0] <= 0.25 and freqs[1] <= 0.25]\n",
    "infrequent_ambigous_freqs = [freqs for freqs in ambigous_freqs.values() if freqs[0] <= 0.25 and freqs[1] <= 0.25]\n",
    "infrequent_likely_pathogenic_freqs = [freqs for freqs in likely_pathogenic_freqs.values() if freqs[0] <= 0.25 and freqs[1] <= 0.25]\n",
    "ax.scatter(*np.array(infrequent_likely_benign_freqs).T, marker='.', color='tab:gray', label=\"likely benign\", zorder=1)\n",
    "ax.scatter(*np.array(infrequent_ambigous_freqs).T, marker='.', color='tab:orange', label=\"ambigous\", zorder=10)\n",
    "ax.scatter(*np.array(infrequent_likely_pathogenic_freqs).T, marker='.', color='tab:red', label=\"likely pathogenic\", zorder=100)\n",
    "\n",
    "notables = []\n",
    "for k in likely_pathogenic_keys + ambigous_keys:\n",
    "    freqs = (jpn_vars[k].snv.AF if k in jpn_vars.keys() else 0, global_vars[k].snv.AF if k in global_vars.keys() else 0)\n",
    "    if freqs[0] <= 0.25 and freqs[1] <= 0.25:\n",
    "        if 0.1 < freqs[0] or 0.1 < freqs[1]:\n",
    "            anno: ensembl.Annotation = jpn_vars.get(k, global_vars[k])\n",
    "            display_name = k[0]\n",
    "            superscript = anno.generic_number if anno.generic_number else anno.segment\n",
    "            substitution = \"{}{}$^{{{}}}${}\".format(anno.ref_aa, anno.residue_number, superscript, anno.alt_aa)\n",
    "            notables.append([display_name, substitution, anno.snv.rsid, anno.pathogenicity, *freqs])\n",
    "notables.sort(key=lambda n: n[5], reverse=True)\n",
    "with open(\"./figures/S4a_notables.csv\", 'w') as f:\n",
    "    f.write('\\n'.join([','.join([str(v) for v in l]) for l in notables]))\n",
    "\n",
    "ax.legend(loc='lower center', bbox_to_anchor=(0.5, 1), ncol=3)\n",
    "lims = (-0.01, 0.26)\n",
    "ticks = [0, 0.05, 0.1, 0.15, 0.2, 0.25]\n",
    "ax.set_xlim(*lims)\n",
    "ax.set_xticks(ticks)\n",
    "ax.set_xticklabels(ticks)\n",
    "ax.set_yticks(ticks)\n",
    "ax.set_yticklabels(ticks)\n",
    "ax.set_ylim(*lims)\n",
    "ax.set_aspect('equal')\n",
    "ax.set_xlabel(\"Allele Freq. in 54KJPN\")\n",
    "ax.set_ylabel(\"Allele Freq. in 1KGP\")\n",
    "fig.tight_layout()\n",
    "fig.savefig(\"./figures/S4a_jpn_vs_global.png\")\n",
    "plt.close(fig)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Fig. S4c\n",
    "fig, ax = plt.subplots(1, 1, figsize=(4, 3), dpi=300)\n",
    "\n",
    "ax.axhspan(AM_THRESHOLD_BENIGN, AM_THRESHOLD_PATHOGENIC, color='lightgray')\n",
    "\n",
    "xys = [(anno.snv.AF, anno.pathogenicity) for anno in jpn_vars.values()]\n",
    "ax.scatter(*np.array(xys).T, marker='.', color='tab:gray')\n",
    "\n",
    "benign_text = \"likely benign\\n({:,})\".format(len(likely_benign_keys))\n",
    "ax.text(1.07, AM_THRESHOLD_BENIGN / 2, benign_text, va='center', ha='center', rotation=90, color='tab:gray', size=8)\n",
    "ambigous_text = \"ambigous\\n({:,})\".format(len(ambigous_keys))\n",
    "ax.text(1.07, (AM_THRESHOLD_BENIGN + AM_THRESHOLD_PATHOGENIC) / 2, ambigous_text, va='center', ha='center', rotation=90, color='tab:gray', size=8)\n",
    "pathogenic_text = \"likely pathogenic\\n({:,} vars.)\".format(len(likely_pathogenic_keys))\n",
    "ax.text(1.07, (AM_THRESHOLD_PATHOGENIC + 1) / 2, pathogenic_text, va='center', ha='center', rotation=90, color='tab:gray', size=8)\n",
    "\n",
    "ax.set_xlabel(\"Allel Freq. in 54KJPN\")\n",
    "ax.set_xlim(-0.02, 1.12)\n",
    "ax.set_ylabel(\"Pathogenicity\")\n",
    "ax.set_ylim(-0.02, 1.02)\n",
    "ax.set_yticks([0, AM_THRESHOLD_BENIGN, AM_THRESHOLD_PATHOGENIC, 1])\n",
    "ax.set_yticklabels([0, AM_THRESHOLD_BENIGN, AM_THRESHOLD_PATHOGENIC, 1])\n",
    "\n",
    "fig.tight_layout()\n",
    "fig.savefig(\"./figures/S4c_freq_vs_pathogenicity.png\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
