{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import gpcrdb\n",
    "import matplotlib\n",
    "matplotlib.use('Agg')\n",
    "matplotlib.rc('pdf', fonttype=42)\n",
    "import matplotlib.pyplot as plt\n",
    "plt.rcParams['font.family'] = \"Arial\"\n",
    "import ensembl\n",
    "from utils import VariationType, Segment\n",
    "import json\n",
    "from scipy import stats\n",
    "import pprint"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "num_high_frequent_missense = {s: 0 for s in Segment} # Fig. 2a\n",
    "\n",
    "names_high_frequent_missense = []\n",
    "annos_high_frequent_missense = []\n",
    "is_A_high_frequent_missense = []\n",
    "seg_length_with_high_frequent_missense = {s: 0 for s in Segment}\n",
    "\n",
    "for receptor in gpcrdb.get_filtered_receptor_list():\n",
    "    ensembl_entry = ensembl.EnsemblGeneEntry(receptor)\n",
    "    with open(receptor.ensembl_path) as f:\n",
    "        display_name = json.load(f)['display_name']\n",
    "    with open(receptor.japan_cds_csv_path) as f:\n",
    "        num_missense = 0\n",
    "\n",
    "        for l in f.readlines():\n",
    "            try:\n",
    "                anno = ensembl.Annotation.from_csv_line(l)\n",
    "                if anno.segment == Segment.FailedToGuess:\n",
    "                    if anno.snv.rsid in ('rs376593544', 'rs1249120368'):\n",
    "                        # Alignment-related issues.\n",
    "                        # These variations are located between H8 and C-term in CCR2.\n",
    "                        # (See `Class A (Rhodopsin)/ccr2_human/alignment.csv`)\n",
    "                        # We reagard these variations as C-term for the following analysis.\n",
    "                        anno.segment = Segment.Cterm\n",
    "                    else:\n",
    "                        raise NotImplementedError\n",
    "                elif anno.segment == Segment.NONE:\n",
    "                    raise NotImplementedError\n",
    "\n",
    "                if anno.var_type == VariationType.MISSENSE and anno.snv.AF > 0.5:\n",
    "                    num_missense += 1\n",
    "\n",
    "                    num_high_frequent_missense[anno.segment] += 1\n",
    "\n",
    "                    names_high_frequent_missense.append(display_name)\n",
    "                    annos_high_frequent_missense.append(anno)\n",
    "                    is_A_high_frequent_missense.append(receptor.receptor_class == 'Class A (Rhodopsin)')\n",
    "            except ensembl.BlankLineError:\n",
    "                continue\n",
    "        if num_missense > 0:\n",
    "            for s in Segment:\n",
    "                seg_length_with_high_frequent_missense[s] += ensembl_entry.segments.count(s)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "meta NOT subset; don't know how to subset; dropped\n"
     ]
    }
   ],
   "source": [
    "# Fig. 2a\n",
    "fig, ax = plt.subplots(1, 1, figsize=(4, 1.5), dpi=300)\n",
    "ax.set_facecolor('whitesmoke')\n",
    "\n",
    "left = 0\n",
    "for s in Segment:\n",
    "    width = num_high_frequent_missense[s]\n",
    "    if width == 0:\n",
    "        continue\n",
    "\n",
    "    ax.barh(0, width, height=0.7, left=left, color=s.color, edgecolor='k', lw=0.2)\n",
    "    \n",
    "    label = s.value\n",
    "    x = left + width / 2\n",
    "    if label.startswith('TM') or label.startswith('H8') or label.endswith('-term'):\n",
    "        if label in ('TM1', 'N-term', 'H8', 'C-term'):\n",
    "            ax.text(x, 0.375, label, ha='center', va='baseline', size=8)\n",
    "        else:\n",
    "            ax.text(x, 0.375, label[-1], ha='center', va='baseline', size=8)\n",
    "\n",
    "    text = str(width) if s != Segment.Nterm else \"{} variants\".format(width)\n",
    "    ax.text(x, -0.375, text, color='black', ha='center', va='top', size=8)\n",
    "\n",
    "    left += width\n",
    "\n",
    "\n",
    "ax.set_xlim(0, left)\n",
    "ax.set_ylim(-0.5, 0.5)\n",
    "ax.set_yticks([])\n",
    "ax.set_yticklabels([])\n",
    "ax.set_xlabel(\"Number of Missense Variants (AF > 0.5)\")\n",
    "fig.tight_layout()\n",
    "fig.savefig(\"./figures/2a_high_allele_freq_vars.pdf\")\n",
    "plt.close(fig)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "M = 54358, n = 23093, N = 120, k = 42\n",
      "N-term hypergemometric test p-value: 0.057480383009668234\n"
     ]
    }
   ],
   "source": [
    "obs_N = num_high_frequent_missense[Segment.Nterm]\n",
    "obs_non_N = sum(num_high_frequent_missense.values()) - obs_N\n",
    "\n",
    "len_N = seg_length_with_high_frequent_missense[Segment.Nterm]\n",
    "len_non_N = sum(seg_length_with_high_frequent_missense.values()) - len_N\n",
    "\n",
    "M = len_N + len_non_N\n",
    "n = len_N\n",
    "N = obs_N + obs_non_N\n",
    "k = obs_N\n",
    "\n",
    "print(f\"M = {M}, n = {n}, N = {N}, k = {k}\")\n",
    "p_hypergeom = stats.hypergeom.cdf(M=M, n=n, N=N, k=k)\n",
    "print(\"N-term hypergemometric test p-value:\", p_hypergeom)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "meta NOT subset; don't know how to subset; dropped\n"
     ]
    }
   ],
   "source": [
    "# Fig. 2b\n",
    "fig, ax = plt.subplots(1, 1, figsize=(4, 12), dpi=300)\n",
    "ax.set_facecolor('whitesmoke')\n",
    "\n",
    "sorted = [(name, anno, is_A) for name, anno, is_A in zip(names_high_frequent_missense, annos_high_frequent_missense, is_A_high_frequent_missense)]\n",
    "sorted.sort(key=lambda arr: (arr[1].snv.AF, -arr[1].segment.index, arr[2], arr[1].residue_number))\n",
    "\n",
    "A_ticks, A_ticklabels = [], []\n",
    "non_A_ticks, non_A_ticklabels = [], []\n",
    "for y, d in enumerate(sorted):\n",
    "    name, anno, is_A = d[0], d[1], d[2]\n",
    "\n",
    "    ax.barh(y, anno.snv.AF, color=anno.segment.color)\n",
    "\n",
    "    if anno.snv.AF == 1:\n",
    "        text = \"1\"\n",
    "    else:\n",
    "        text = \"> {:.2f}\".format(int(anno.snv.AF * 100) / 100)\n",
    "    ax.text(anno.snv.AF + 0.01, y, text, size=5, ha='left', va='center')\n",
    "\n",
    "    label = \"{} {}{}{} ({})\".format(name, anno.ref_aa, anno.residue_number, anno.alt_aa, \n",
    "                                    anno.generic_number if anno.generic_number else anno.segment.value)\n",
    "    if is_A:\n",
    "        A_ticks.append(y)\n",
    "        A_ticklabels.append(label)\n",
    "    else:\n",
    "        non_A_ticks.append(y)\n",
    "        non_A_ticklabels.append(label)\n",
    "\n",
    "ax.set_yticks(A_ticks)\n",
    "ax.set_yticklabels(A_ticklabels, size=7)\n",
    "ax.set_yticks(non_A_ticks, minor=True)\n",
    "ax.set_yticklabels(non_A_ticklabels, size=7, minor=True, color='tab:gray')\n",
    "ax.set_xlabel(\"Allele Freq.\")\n",
    "ax.set_ylim(-0.8, len(sorted) - 0.3)\n",
    "ax.set_xticks([0.5, 0.6, 0.8, 1.0])\n",
    "ax.set_xticklabels([0.5, 0.6, 0.8, 1.0])\n",
    "ax.set_xlim(left=0.5, right=1.1)\n",
    "fig.tight_layout()\n",
    "fig.savefig(\"./figures/2b_high_allele_freq_vars.pdf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "def _is_N_glycosylation_motif(triplet):\n",
    "    assert(len(triplet) == 3)\n",
    "\n",
    "    # Check if the variation is related with [N]-X-S/T motif, where X is not Proline.\n",
    "    if triplet[0] == 'N' and triplet[1] != 'P' and triplet[2] in 'ST':\n",
    "        return True\n",
    "    return False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "n_glyco_gain, n_glyco_loss = [], []\n",
    "o_glyco_gain, o_glyco_loss = [], []\n",
    "phospho_gain, phospho_loss = [], []\n",
    "for receptor in gpcrdb.get_filtered_receptor_list():\n",
    "    with open(receptor.ensembl_path) as f:\n",
    "        display_name = json.load(f)['display_name']\n",
    "    with open(receptor.alignment_path) as f:\n",
    "        seq = ''.join([l.split(',')[0][0] for l in f if not l.startswith('#')])\n",
    "    with open(receptor.japan_cds_csv_path) as f:\n",
    "        for l in f:\n",
    "            try:\n",
    "                anno = ensembl.Annotation.from_csv_line(l)  \n",
    "            except ensembl.BlankLineError:\n",
    "                continue\n",
    "            \n",
    "            if anno.var_type != VariationType.MISSENSE:\n",
    "                continue\n",
    "            if anno.snv.AF <= 0.5:\n",
    "                continue\n",
    "            assert(seq[anno.residue_number - 1] == anno.ref_aa)\n",
    "            \n",
    "            # Check glycosylation\n",
    "            if anno.segment in (Segment.Nterm, Segment.ECL1, Segment.ECL2, Segment.ECL3):    \n",
    "                # N-glycosylation\n",
    "                if 0 <= anno.residue_number - 3 and anno.residue_number <= len(seq):\n",
    "                    ref_leading_triplet = seq[anno.residue_number - 3:anno.residue_number]\n",
    "                    alt_leading_triplet = ref_leading_triplet[0] + ref_leading_triplet[1]+ anno.alt_aa\n",
    "\n",
    "                    if _is_N_glycosylation_motif(ref_leading_triplet) != _is_N_glycosylation_motif(alt_leading_triplet):\n",
    "                        if _is_N_glycosylation_motif(ref_leading_triplet):\n",
    "                            # Loss\n",
    "                            n_glyco_loss.append({\"display_name\": display_name, \"annotation\": anno, \n",
    "                                                \"ref_motif\": ref_leading_triplet, \"alt_motif\": alt_leading_triplet})\n",
    "                        else:\n",
    "                            # Gain\n",
    "                            n_glyco_gain.append({\"display_name\": display_name, \"annotation\": anno,\n",
    "                                                \"ref_motif\": ref_leading_triplet, \"alt_motif\": alt_leading_triplet})\n",
    "\n",
    "                if 0 <= anno.residue_number - 2 and anno.residue_number + 1 <= len(seq):\n",
    "                    ref_midmost_triplet = seq[anno.residue_number - 2:anno.residue_number + 1]\n",
    "                    alt_midmost_triplet = ref_midmost_triplet[0] + anno.alt_aa + ref_midmost_triplet[2]\n",
    "\n",
    "                    if _is_N_glycosylation_motif(ref_midmost_triplet) != _is_N_glycosylation_motif(alt_midmost_triplet):\n",
    "                        if _is_N_glycosylation_motif(ref_midmost_triplet):\n",
    "                            # Loss\n",
    "                            n_glyco_loss.append({\"display_name\": display_name, \"annotation\": anno,\n",
    "                                                \"ref_motif\": ref_midmost_triplet, \"alt_motif\": alt_midmost_triplet})\n",
    "                        else:\n",
    "                            # Gain\n",
    "                            n_glyco_gain.append({\"display_name\": display_name, \"annotation\": anno,\n",
    "                                                \"ref_motif\": ref_midmost_triplet, \"alt_motif\": alt_midmost_triplet})\n",
    "\n",
    "                if 0 <= anno.residue_number - 1 and anno.residue_number + 2 <= len(seq):\n",
    "                    ref_tailing_triplet = seq[anno.residue_number - 1:anno.residue_number + 2]\n",
    "                    alt_tailing_triplet = anno.alt_aa + ref_tailing_triplet[1] + ref_tailing_triplet[2]\n",
    "\n",
    "                    if _is_N_glycosylation_motif(ref_tailing_triplet) != _is_N_glycosylation_motif(alt_tailing_triplet):\n",
    "                        if _is_N_glycosylation_motif(ref_tailing_triplet):\n",
    "                            # Loss\n",
    "                            n_glyco_loss.append({\"display_name\": display_name, \"annotation\": anno,\n",
    "                                                \"ref_motif\": ref_tailing_triplet, \"alt_motif\": alt_tailing_triplet})\n",
    "                        else:\n",
    "                            # Gain\n",
    "                            n_glyco_gain.append({\"display_name\": display_name, \"annotation\": anno,\n",
    "                                                \"ref_motif\": ref_tailing_triplet, \"alt_motif\": alt_tailing_triplet})\n",
    "                # O-glycosylation\n",
    "                if anno.ref_aa in 'ST' and anno.alt_aa not in 'ST':\n",
    "                    # Loss\n",
    "                    o_glyco_loss.append({\"display_name\": display_name, \"annotation\": anno})\n",
    "                elif anno.ref_aa not in 'ST' and anno.alt_aa in 'ST':\n",
    "                    # Gain\n",
    "                    o_glyco_gain.append({\"display_name\": display_name, \"annotation\": anno})\n",
    "            # Check phosphorylation\n",
    "            if anno.segment in (Segment.Cterm, Segment.ICL1, Segment.ICL2, Segment.ICL3, Segment.ICL4):\n",
    "                if anno.ref_aa in 'ST' and anno.alt_aa not in 'ST':\n",
    "                    # Loss\n",
    "                    phospho_loss.append({\"display_name\": display_name, \"annotation\": anno})\n",
    "                elif anno.ref_aa not in 'ST' and anno.alt_aa in 'ST':\n",
    "                    # Gain\n",
    "                    phospho_gain.append({\"display_name\": display_name, \"annotation\": anno})\n",
    "n_glyco_gain.sort(key=lambda d: d['annotation'].snv.AF)\n",
    "n_glyco_loss.sort(key=lambda d: d['annotation'].snv.AF)\n",
    "o_glyco_gain.sort(key=lambda d: d['annotation'].snv.AF)\n",
    "o_glyco_loss.sort(key=lambda d: d['annotation'].snv.AF)\n",
    "phospho_gain.sort(key=lambda d: d['annotation'].snv.AF)\n",
    "phospho_loss.sort(key=lambda d: d['annotation'].snv.AF)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "M = 9104, n = 7178, N = 7, k = 7\n",
      "Phosphorylation hypergemometric test p-value: 1.0\n"
     ]
    }
   ],
   "source": [
    "len_C = seg_length_with_high_frequent_missense[Segment.Cterm]\n",
    "len_ICLs = sum([seg_length_with_high_frequent_missense[s] for s in Segment.ICLs()])\n",
    "\n",
    "obs_C, obs_ICLs = 0, 0\n",
    "for p in phospho_gain + phospho_loss:\n",
    "    if p['annotation'].segment == Segment.Cterm:\n",
    "        obs_C += 1\n",
    "    if p['annotation'].segment in Segment.ICLs():\n",
    "        obs_ICLs += 1\n",
    "\n",
    "M = len_C + len_ICLs\n",
    "n = len_C\n",
    "N = obs_C + obs_ICLs\n",
    "k = obs_C\n",
    "\n",
    "print(f\"M = {M}, n = {n}, N = {N}, k = {k}\")\n",
    "p_hypergeom = stats.hypergeom.cdf(M=M, n=n, N=N, k=k)\n",
    "print(\"Phosphorylation hypergemometric test p-value:\", p_hypergeom)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "meta NOT subset; don't know how to subset; dropped\n"
     ]
    }
   ],
   "source": [
    "fig, axes = plt.subplots(3, 1, figsize=(4, 6), dpi=300, sharex=True, sharey=True)\n",
    "for ax in axes:\n",
    "    ax.set_facecolor('whitesmoke')\n",
    "    ax.set_yticks([0.5, 0.6, 0.7, 0.8, 0.9, 1])\n",
    "    ax.set_yticklabels([0.5, 0.6, 0.7, 0.8, 0.9, 1])\n",
    "    ax.set_ylabel(\"Allele Freq.\")\n",
    "\n",
    "ax = axes[0]\n",
    "n_glyco = [[d['annotation'].snv.AF for d in n_glyco_gain], [d['annotation'].snv.AF for d in n_glyco_loss]]\n",
    "ax.hist(n_glyco, bins=10, range=(0.5, 1), stacked=True, label=[\"Gain\", \"Loss\"], orientation='horizontal',\n",
    "        color=['tab:orange', 'tab:gray'])\n",
    "ax.set_title(\"N-glycosylation (Asn-X-Ser/Thr)\") \n",
    "for glyco in n_glyco_gain + n_glyco_loss:\n",
    "    anno = glyco['annotation']\n",
    "    from_to_label = \"({}$\\\\rightarrow${})\".format(glyco['ref_motif'], glyco['alt_motif'])\n",
    "    superscript = anno.generic_number if anno.generic_number else anno.segment.value\n",
    "    superscript = r'$^{\\mathrm{' + superscript + r'}}$'\n",
    "    label = \"{} {}{}{}{} {}\".format(glyco['display_name'], anno.ref_aa, anno.residue_number, superscript, anno.alt_aa, from_to_label)\n",
    "    for r in range(50, 100, 5):\n",
    "        bottom, top = r / 100, r / 100 + 0.05\n",
    "        if bottom < anno.snv.AF <= top:\n",
    "            ax.text(1, (bottom + top) / 2 - 0.005, \" \" + label, ha='left', va='center', size=9)\n",
    "\n",
    "ax = axes[1]\n",
    "o_glyco = [[d['annotation'].snv.AF for d in o_glyco_gain], [d['annotation'].snv.AF for d in o_glyco_loss]]\n",
    "ax.hist(o_glyco, bins=10, range=(0.5, 1), stacked=True, label=[\"Gain\", \"Loss\"], orientation='horizontal',\n",
    "        color=['tab:orange', 'tab:gray'])\n",
    "ax.set_title(\"O-glycosylation (Ser/Thr)\")\n",
    "\n",
    "ax = axes[2]\n",
    "phospho = [[d['annotation'].snv.AF for d in phospho_gain], [d['annotation'].snv.AF for d in phospho_loss]]\n",
    "ax.hist(phospho, bins=10, range=(0.5, 1), stacked=True, label=[\"Gain\", \"Loss\"], orientation='horizontal',\n",
    "        color=['tab:orange', 'tab:gray'])\n",
    "ax.set_title(\"Phosphorylation (Ser/Thr)\")\n",
    "ax.legend(loc='lower right', ncol=2)\n",
    "ax.set_xticks(range(6))\n",
    "ax.set_xticklabels(range(6))\n",
    "ax.set_xlabel(\"Number of Variants\")\n",
    "\n",
    "fig.tight_layout()    \n",
    "fig.savefig(\"./figures/2cde_ptm.pdf\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
